crates/sound/src/lib.rs

```diff
@@ -1,73 +1,196 @@
 pub type Result<T> = std::result::Result<T, Error>;
 
 /// Custom error types
 #[derive(Debug, ThisError)]
 pub enum Error {
+    #[error("Notification send failed")]
+    SendNotificationFailed,
+    #[error("Descriptor not found")]
+    DescriptorNotFound,
+    #[error("Descriptor read failed")]
+    DescriptorReadFailed,
+    #[error("Descriptor write failed")]
+    DescriptorWriteFailed,
     #[error("Failed to handle event other than EPOLLIN event")]
     HandleEventNotEpollIn,
     #[error("Failed to handle unknown event")]
     HandleUnknownEvent,
+    #[error("Invalid control message code {0}")]
+    InvalidControlMessage(u32),
     #[error("Failed to create a new EventFd")]
     EventFdCreate(IoError),
     #[error("Request missing data buffer")]
     SoundReqMissingData,
     #[error("Audio backend not supported")]
     AudioBackendNotSupported,
-    #[error("Descriptor not found")]
-    DescriptorNotFound,
-    #[error("Descriptor read failed")]
-    DescriptorReadFailed,
-    #[error("Descriptor write failed")]
-    DescriptorWriteFailed,
-    #[error("Isufficient descriptor size, required: {0}, found: {1}")]
-    InsufficientDescriptorSize(usize, usize),
-    #[error("Failed to send notification")]
-    SendNotificationFailed,
+    #[error("Invalid virtio_snd_hdr size, expected: {0}, found: {1}")]
+    UnexpectedSoundHeaderSize(usize, u32),
+    #[error("Received unexpected write only descriptor at index {0}")]
+    UnexpectedWriteOnlyDescriptor(usize),
+    #[error("Received unexpected readable descriptor at index {0}")]
+    UnexpectedReadableDescriptor(usize),
     #[error("Invalid descriptor count {0}")]
     UnexpectedDescriptorCount(usize),
     #[error("Invalid descriptor size, expected: {0}, found: {1}")]
-    UnexpectedDescriptorSize(usize, usize),
-    #[error("Invalid descriptor size, expected at least: {0}, found: {1}")]
-    UnexpectedMinimumDescriptorSize(usize, usize),
-    #[error("Received unexpected readable descriptor at index {0}")]
-    UnexpectedReadableDescriptor(usize),
-    #[error("Received unexpected write only descriptor at index {0}")]
-    UnexpectedWriteOnlyDescriptor(usize),
+    UnexpectedDescriptorSize(usize, u32),
+    #[error("Protocol or device error: {0}")]
+    Stream(stream::Error),
 }
 
-impl std::convert::From<Error> for IoError {
+impl From<Error> for IoError {
     fn from(e: Error) -> Self {
-        IoError::new(ErrorKind::Other, e)
+        Self::new(ErrorKind::Other, e)
     }
 }
 
-#[derive(Default, Clone)]
-pub struct PCMParams {
-    pub features: Le32,
-    /// size of hardware buffer in bytes
-    pub buffer_bytes: Le32,
-    /// size of hardware period in bytes
-    pub period_bytes: Le32,
-    pub channels: u8,
-    pub format: u8,
-    pub rate: u8,
+impl From<stream::Error> for Error {
+    fn from(val: stream::Error) -> Self {
+        Self::Stream(val)
+    }
+}
+
+#[derive(Debug)]
+pub struct InvalidControlMessage(u32);
```

Comment:
needs a std::error::Error impl as well

Reply:
Anything that implements `std::fmt::{Debug,Display}` also implements `Error`. The `Error` trait has only provided methods, meaning you would only manually implement it to override the default methods.

Reply:
That is wrong.

Reply:
The Error trait required Debug + Display, so there can't be a blanked implementation for that.

Reply:
Thanks, I misremembered. I will add an impl in a fixup commit

Reply:
fixed